### PR TITLE
Add a test service for the W3C validation harness

### PR DESCRIPTION
**Pull request by pyohannes**
This example program is intended to be used as a test service for the [W3C Distributed Tracing Validation Service](https://github.com/w3c/trace-context/tree/master/test). It is implemented according to [this instructions](https://github.com/w3c/trace-context/tree/master/test#implement-test-service).

With the current state of our implementation, 26 out of 40 tests pass. Most failures are related to `tracestate` propagation, which we currently do not support at all.

This program makes use of `nlohmann::json` and our internally provided HTTP client and server implementations.

### CODE REVIEW

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -6,3 +6,4 @@ add_subdirectory(simple)
 add_subdirectory(batch)
 add_subdirectory(metrics_simple)
 add_subdirectory(multithreaded)
+add_subdirectory(w3c_tracecontext_test)
```
Comment (ThomsonTan) [inline examples/CMakeLists.txt:9]: Put w3c_tracecontext_test into some test folder instead of under examples?

Comment (pyohannes) [reply examples/CMakeLists.txt:9]: Yes, I'll move it. Probably I'll set up something similar to what C# people did: https://github.com/open-telemetry/opentelemetry-dotnet/tree/master/test/OpenTelemetry.Instrumentation.W3cTraceContext.Tests

Comment (lalitb) [reply examples/CMakeLists.txt:9]: Yes, docker images or script to run both the services and give result would be handy ( as part of separate PR )

File: examples/w3c_tracecontext_test/main.cc [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+  Uri(std::string uri)
+  {
+    size_t host_end = uri.substr(7, std::string::npos).find(":");
+    size_t port_end = uri.substr(host_end + 1, std::string::npos).find("/");
+
+    host = uri.substr(0, host_end + 7);
+    port = std::stoi(uri.substr(7 + host_end + 1, port_end));
+    path = uri.substr(host_end + port_end + 2, std::string::npos);
+  }
+};
```
Comment (lalitb) [inline examples/w3c_tracecontext_test/main.cc:70]: jfi - i added a url_parser.h header to parse the url as part of zipkin exporter PR  ( https://github.com/open-telemetry/opentelemetry-cpp/pull/471/files )

File: examples/CMakeLists.txt (hunk 1 above)
Comment (lalitb) [reply examples/CMakeLists.txt:9]: Right, would be good to have a script / docker image to run both the services and give the result.

**Review by lalitb — APPROVED** [review]
LGTM. Thanks for this adding validation code.

> Most failures are related to tracestate propagation, which we currently do not support at all.

Should we be creating a ticket for this ?

File: examples/CMakeLists.txt (hunk 1 above)
Comment (lalitb) [reply examples/CMakeLists.txt:9]: Yes, docker images or script to run both the services and give result would be handy ( as part of separate PR )

File: examples/w3c_tracecontext_test/main.cc [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+#include "opentelemetry/ext/http/client/curl/http_client_curl.h"
+#include "opentelemetry/ext/http/server/http_server.h"
+#include "opentelemetry/sdk/trace/simple_processor.h"
+#include "opentelemetry/sdk/trace/tracer_provider.h"
+#include "opentelemetry/trace/propagation/http_trace_context.h"
+#include "opentelemetry/trace/provider.h"
+#include "opentelemetry/trace/scope.h"
+
+#include <algorithm>
+#include "nlohmann/json.hpp"
```
Comment (maxgolov) [inline examples/w3c_tracecontext_test/main.cc:12]: Where / how do you install `nlohmann/json.hpp` for non-Windows OS?

File: examples/CMakeLists.txt (hunk 1 above)
Comment (pyohannes) [reply examples/CMakeLists.txt]: This is moved now into `ext`, signaling that this is not a central component or test that has to run with the core test suite.

File: examples/w3c_tracecontext_test/main.cc (hunk 2 above)
Comment (pyohannes) [reply examples/w3c_tracecontext_test/main.cc]: I'd love to use that once it's merged.

File: examples/w3c_tracecontext_test/main.cc (hunk 3 above)
Comment (pyohannes) [reply examples/w3c_tracecontext_test/main.cc]: This relies on the `nlohmann-json` packages installed by our CI tools.

**Comment by pyohannes** [comment]
This should be ready for another review and hopefully merge, all the CI tests are passing.

**Review by ThomsonTan — APPROVED** [review]

**Pull request merged**